File: packages/froxel-math/src/vec2.ts

import { Module } from "asyl";

let m: Module = null;

/**
 * 2D Vector.
 */
export default class Vec2
{
	/**
	 * Address of underlying vec2.
	 */
	addr: number;

	/**
	 * Underlying vector data.
	 */
	data: Float32Array;

	/**
	 * Binds the specified Asyl module to the class.
	 */
	static bind (module: Module) : void {
		m = module;
	}

	/**
	 * Allocates a new vector with the specified coordinates.
	 */
	static alloc (x:number=0, y:number=0) : Vec2 {
		return new Vec2(m.vec2_alloc2f(x, y));
	}

	/**
	 * Constructs the vector with the specified coordinates.
	 */
	private constructor (addr: number) {
		this.addr = addr;
		this.data = m.mapFloat32Array(addr, 2);
	}

	/**
	 * Destroys the instance.
	 */
	dtor() {
		m.vec2_dtor(this.addr);
	}

	/**
	 * Returns a new vector with the same coordinates.
	 */
	clone() : Vec2 {
		return new Vec2(m.vec2_clone(this.addr));
	}

	/**
	 * Sets the coordinates of the vector from a Vec2 object.
	 */
	set (v: Vec2) : Vec2;
	/**
	 * Sets the coordinates of the vector.
	 */
	set (x: number, y: number) : Vec2;

	set (x: Vec2|number, y: number = null) : Vec2
	{
		if (y === null)
			m.vec2_set1v(this.addr, (x as Vec2).addr);
		else
			m.vec2_set2f(this.addr, x, y);

		return this;
	}

	/**
	 * Sets the X-coordinate of the vector.
	 */
	setX (x: number) : Vec2 {
		this.data[0] = x;
		return this;
	}

	/**
	 * Sets the Y-coordinate of the vector.
	 */
	setY (y: number) : Vec2 {
		this.data[1] = y;
		return this;
	}

	/**
	 * Returns the value of the X-coordinate.
	 */
	x() : number {
		return this.data[0];
	}

	/**
	 * Returns the value of the Y-coordinate.
	 */
	y() : number {
		return this.data[1];
	}

	/**
	 * Sets the coordinates of the vector to zero.
	 */
	zero() : Vec2 {
		m.vec2_zero(this.addr);
		return this;
	}

	/**
	 * Returns true if the vector coordinates are zero.
	 */
	iszero() : boolean {
		return m.vec2_iszero(this.addr);
	}

	/**
	 * Returns true if the coordinates of the vector have the same values as the given Vec2.
	 */
	equals (v: Vec2) : boolean;
	/**
	 * Returns true if the coordinates of the vector have the same values as the given ones.
	 */
	equals (x: number, y: number) : boolean;

	equals (x: Vec2|number, y: number=null) : boolean
	{
		if (y === null)
			return m.vec2_equals1v(this.addr, (x as Vec2).addr);

		return m.vec2_equals2f(this.addr, x, y);
	}

	/**
	 * Returns true if the difference between the coordinates is less than the specified epsilon.
	 */
	almost (v: Vec2, epsilon: number) : boolean;
	/**
	 * Returns true if the difference between the coordinates is less than the specified epsilon.
	 */
	almost (x: number, y: number, epsilon: number) : boolean;

	almost (x: Vec2|number, y: number=null, epsilon: number=null) : boolean
	{
		if (epsilon === null)
			return m.vec2_almost1v(this.addr, (x as Vec2).addr, y);

		return m.vec2_almost2f(this.addr, x, y, epsilon);
	}

	/**
	 * Negates the vector, that is changing the sign of each component in the vector.
	 */
	neg() : Vec2 {
		m.vec2_neg(this.addr);
		return this;
	}

	/**
	 * Inverts the vector by changing each component to its reciprocal.
	 */
	inv() : Vec2 {
		m.vec2_inv(this.addr);
		return this;
	}

	/**
	 * Changes the components of the vector to their absolute value.
	 */
	abs() : Vec2 {
		m.vec2_abs(this.addr);
		return this;
	}

	/**
	 * Adds the coordinates of the given Vec2 to the vector.
	 */
	translate (v: Vec2) : Vec2;
	/**
	 * Adds the given delta values to the vector.
	 */
	translate (dx: number, dy: number) : Vec2;

	translate (dx: Vec2|number, dy: number=null) : Vec2
	{
		if (dy === null)
			m.vec2_translate1v(this.addr, (dx as Vec2).addr);
		else
			m.vec2_translate2f(this.addr, dx, dy);

		return this;
	}

	/**
	 * Rotates the vector by the specified angle (in radians) using the specified (optional) origin point.
	 */
	rotate (angle: number, cx: number=0, cy: number=0) : Vec2
	{
		m.vec2_rotate3f(this.addr, angle, cx, cy);
		return this;
	}

	/**
	 * Adds the coordinates of the given Vec2 to the vector.
	 */
	add (v: Vec2) : Vec2;
	/**
	 * Adds the given delta values to the vector.
	 */
	add (dx: number, dy: number) : Vec2;

	add (dx: Vec2|number, dy: number=null)
	{
		if (dy === null)
			m.vec2_add1v(this.addr, (dx as Vec2).addr);
		else
			m.vec2_add2f(this.addr, dx, dy);

		return this;
	}

	/**
	 * Subtracts the coordinates of the given Vec2 from the vector.
	 */
	sub (v: Vec2) : Vec2;
	/**
	 * Subtracts the given delta values from the vector.
	 */
	sub (dx: number, dy: number) : Vec2;

	sub (dx: Vec2|number, dy: number=null)
	{
		if (dy === null)
			m.vec2_sub1v(this.addr, (dx as Vec2).addr);
		else
			m.vec2_sub2f(this.addr, dx, dy);

		return this;
	}

	/**
	 * Scales each components of the vector by the respective component of the given one.
	 */
	scale (v: Vec2) : Vec2;
	/**
	 * Scales both components of the vector by the given factor.
	 */
	scale (factor: number) : Vec2;
	/**
	 * Scales each components of the vector by the given factors.
	 */
	scale (fx: number, fy: number) : Vec2;

	scale (fx: Vec2|number, fy: number=null) : Vec2
	{
		if (fy === null)
		{
			if (fx instanceof Vec2)
				m.vec2_scale1v(this.addr, fx.addr);
			else
				m.vec2_scale1f(this.addr, fx);
		}
		else
			m.vec2_scale2f(this.addr, fx, fy);

		return this;
	}

	/**
	 * Performs `floor` on the components of the vector.
	 */
	floor() : Vec2 {
		m.vec2_floor(this.addr);
		return this;
	}

	/**
	 * Performs `ceil` on the components of the vector.
	 */
	ceil() : Vec2 {
		m.vec2_ceil(this.addr);
		return this;
	}

	/**
	 * Truncates the vector components to their integer parts.
	 */
	trunc() : Vec2 {
		m.vec2_trunc(this.addr);
		return this;
	}

	/**
	 * Truncates the vector components to their fractional parts.
	 */
	fract() : Vec2 {
		m.vec2_fract(this.addr);
		return this;
	}

	/**
	 * Returns the dot product of the vectors.
	 */
	dot (v: Vec2) : number;
	/**
	 * Returns the dot product of the vector and the given values.
	 */
	dot (x: number, y: number) : number;

	dot (x: Vec2|number, y: number=null) : number
	{
		if (y === null)
			return m.vec2_dot1v(this.addr, (x as Vec2).addr);

		return m.vec2_dot2f(this.addr, x, y);
	}

	/**
	 * Returns the squared magnitude of the vector.
	 */
	mag2() : number {
		return m.vec2_mag2(this.addr);
	}

	/**
	 * Returns the magnitude of the vector.
	 */
	mag() : number {
		return m.vec2_mag(this.addr);
	}

	/**
	 * Normalizes the vector by dividing each component by the vector magnitude to obtain a unit vector.
	 */
	unit() : Vec2 {
		m.vec2_unit(this.addr);
		return this;
	}

	/**
	 * Sets the vector to its major-axis, that is the component with the maximum absolute value.
	 */
	major() : Vec2 {
		m.vec2_major(this.addr);
		return this;
	}

	/**
	 * Sets the vector to its minor-axis, that is the component with the minimum absolute value.
	 */
	minor() : Vec2 {
		m.vec2_minor(this.addr);
		return this;
	}

	/**
	 * Sets the vector to its sign-vector representation.
	 */
	sign() : Vec2 {
		m.vec2_sign(this.addr);
		return this;
	}

	/**
	 * Returns the string representation of the coordinates of the vector.
	 */
	toString() : string
	{
		return `(${this.x()}, ${this.y()})`;
	}
};

File: packages/froxel-math/src/vec2.cpp

#include <cstring>
#include <cstdlib>
#include <cmath>
#include <wasm>

#include "vec2.h"


export vec2 *vec2_alloc2f (double x, double y) {
	vec2 *t = (vec2*)malloc(sizeof(vec2));
	t->x = x;
	t->y = y;
	return t;
}

export vec2 *vec2_materialize (void *addr) {
	return (vec2*)addr;
}

export void vec2_free (vec2 *self) {
	free (self);
}

export vec2 *vec2_clone (vec2 *self) {
	return vec2_alloc2f(self->x, self->y);
}

export void vec2_set2f (vec2 *self, double x, double y) {
	self->x = x;
	self->y = y;
}

export void vec2_set1v (vec2 *self, vec2 *v) {
	self->x = v->x;
	self->y = v->y;
}

export void vec2_zero (vec2 *self) {	
	memset(self, 0, sizeof(vec2));
}

export bool vec2_iszero (vec2 *self) {
	return !self->x && !self->y;
}

export bool vec2_equals2f (vec2 *self, double x, double y) {
	return self->x == x && self->y == y;
}

export bool vec2_equals1v (vec2 *self, vec2 *v) {
	return self->x == v->x && self->y == v->y;
}

export bool vec2_almost2f (vec2 *self, double x, double y, double epsilon) {
	return std::abs(self->x - x) < epsilon && std::abs(self->y - y) < epsilon;
}

export bool vec2_almost1v (vec2 *self, vec2 *v, double epsilon) {
	return std::abs(self->x - v->x) < epsilon && std::abs(self->y - v->y) < epsilon;
}

export void vec2_neg (vec2 *self) {
	self->x = -self->x;
	self->y = -self->y;
}

export void vec2_inv (vec2 *self) {
	self->x = 1.0 / self->x;
	self->y = 1.0 / self->y;
}

export void vec2_abs (vec2 *self) {
	self->x = std::abs(self->x);
	self->y = std::abs(self->y);
}

export void vec2_translate2f (vec2 *self, double dx, double dy) {
	self->x += dx;
	self->y += dy;
}

export void vec2_translate1v (vec2 *self, vec2 *v) {
	self->x += v->x;
	self->y += v->y;

}

export void vec2_rotate3f (vec2 *self, double angle, double cx, double cy)
{
	double cost = std::cos(angle);
	double sint = std::sin(angle);

	self->x -= cx;
	self->y -= cy;

	double x = self->x*cost + self->y*sint;
	double y = self->y*cost - self->x*sint;

	self->x = x + cx;
	self->y = y + cy;
}

export void vec2_add2f (vec2 *self, double dx, double dy) {
	self->x += dx;
	self->y += dy;
}

export void vec2_add1v (vec2 *self, vec2 *v) {
	self->x += v->x;
	self->y += v->y;
}

export void vec2_sub2f (vec2 *self, double dx, double dy) {
	self->x -= dx;
	self->y -= dy;
}

export void vec2_sub1v (vec2 *self, vec2 *v) {
	self->x -= v->x;
	self->y -= v->y;
}

export void vec2_scale1f (vec2 *self, double f) {
	self->x *= f;
	self->y *= f;
}

export void vec2_scale2f (vec2 *self, double fx, double fy) {
	self->x *= fx;
	self->y *= fy;
}

export void vec2_scale1v (vec2 *self, vec2 *v) {
	self->x *= v->x;
	self->y *= v->y;
}

export void vec2_floor (vec2 *self) {
	self->x = std::floor(self->x);
	self->y = std::floor(self->y);
}

export void vec2_ceil (vec2 *self) {
	self->x = std::ceil(self->x);
	self->y = std::ceil(self->y);
}

export void vec2_trunc (vec2 *self) {
	self->x = (int)(self->x);
	self->y = (int)(self->y);
}

export void vec2_fract (vec2 *self) {
	self->x = self->x - (int)(self->x);
	self->y = self->y - (int)(self->y);
}

export double vec2_dot2f (vec2 *self, double x, double y) {
	return self->x*x + self->y*y;
}

export double vec2_dot1v (vec2 *self, vec2 *v) {
	return self->x*v->x + self->y*v->y;
}

export double vec2_mag2 (vec2 *self) {
	return self->x*self->x + self->y*self->y;
}

export double vec2_mag (vec2 *self) {
	return std::sqrt(vec2_mag2(self));
}

export void vec2_unit (vec2 *self) {
	if (vec2_iszero(self)) return;
	vec2_scale1f(self, 1.0 / vec2_mag(self));
}

export void vec2_major (vec2 *self) {
	if (std::abs(self->x) > std::abs(self->y))
		self->y = 0;
	else
		self->x = 0;
}

export void vec2_minor (vec2 *self) {
	if (std::abs(self->x) < std::abs(self->y))
		self->y = 0;
	else
		self->x = 0;
}

export void vec2_sign (vec2 *self) {
	self->x = !self->x ? 0 : (self->x < 0 ? -1 : 1);
	self->y = !self->y ? 0 : (self->y < 0 ? -1 : 1);
}
